### ISSUE TITLE
setInterval and setTimeout not working

### ISSUE THREAD

**Issue by qwe123wsx** [LiquidPlayer/LiquidCore#94]
I tried using this code, setInterval and setTimeout doesnt work.

```javascript
JSContext context = new JSContext();
JSFunction log = new JSFunction(context,"log") {
    public void log(String x) {
         Log.d("Log", "log:" + x);
    }
};
context.property("log", log);
context.evaluateScript("(() => { setInterval(function() {log('test')}, 1000)} )()");
```
Here is the error:
```
2019-04-06 12:20:53.889 15798-15798/id.shortcut.testliquidcore E/AndroidRuntime: FATAL EXCEPTION: main
    Process: id.shortcut.testliquidcore, PID: 15798
    java.lang.RuntimeException: Unable to start activity ComponentInfo{id.shortcut.testliquidcore/id.shortcut.testliquidcore.MainActivity}: ReferenceError: setInterval is not defined
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2913)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: ReferenceError: setInterval is not defined
        at org.liquidplayer.javascript.JSContext.evaluateScript(JSContext.java:203)
        at org.liquidplayer.javascript.JSContext.evaluateScript(JSContext.java:215)
        at id.shortcut.testliquidcore.MainActivity.test(MainActivity.java:22)
        at id.shortcut.testliquidcore.MainActivity.onCreate(MainActivity.java:29)
        at android.app.Activity.performCreate(Activity.java:7136)
        at android.app.Activity.performCreate(Activity.java:7127)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2893)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6669) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
```

How can i make it work?

**Comment by ericwlange** [comment]
Hi, welcome to LiquidCore.

`setTimeout` and `setInterval` are not part of raw Javascript.  In a browser, they are implemented in the DOM.  So if you are using LiquidCore and as raw Javascript engine, you will have to roll your own.  If, however, you use the Node.js VM through a `MicroService`, those functions are available out of the box.

Creating your own implementation is pretty straightforward, though.  Here is an untested example of how you might implement `setTimeout`:

Keep track of timers in a hash map so they can be cancelled later with `clearTimeout()`:
```java
private SparseArray<Thread> threadMap = new SparseArray<>();
```

Then implement `setTimeout`:
```java
context.property("setTimeout", new JSFunction(context, "setTimeout") {
    public Integer setTimeout(final JSFunction fn, final Integer ms, final JSValue ... args) {
        final Thread thread = new Thread() {
            @Override public void run() {
                Thread.sleep(ms);
                if (threadMap.get(thread.hashCode()) == thread) {
                    fn.apply(null, args);
                    threadMap.remove(thread.hashCode());
                }
            }
        };
        threadMap.put(thread.hashCode(), thread);
        thread.start();
        return thread.hashCode();
    }
});
```

`setInterval` would be similar, except the the `run()` routine would operate inside an endless loop that breaks when the hash code is not found in the map anymore.

You could then implement `clearTimeout` by simply removing the hash from the map.

**Comment by ericwlange** [comment]
Complete implementation (again untested, there may be typos):

```javascript
context.property("setTimeout", new JSFunction(context, "setTimeout") {
    public Integer setTimeout(final JSFunction fn, final Integer ms, final JSValue ... args) {
        final Thread thread = new Thread() {
            @Override public void run() {
                Thread.sleep(ms);
                if (threadMap.get(thread.hashCode()) == thread) {
                    fn.apply(null, args);
                    threadMap.remove(thread.hashCode());
                }
            }
        };
        threadMap.put(thread.hashCode(), thread);
        thread.start();
        return thread.hashCode();
    }
});

context.property("setInterval", new JSFunction(context, "setInterval") {
    public Integer setInterval(final JSFunction fn, final Integer ms, final JSValue ... args) {
        final Thread thread = new Thread() {
            @Override public void run() {
                while (true) {
                    Thread.sleep(ms);
                    if (threadMap.get(thread.hashCode()) == thread) {
                        fn.apply(null, args);
                        threadMap.remove(thread.hashCode());
                    } else (
                        break;
                    }
                }
            }
        };
        threadMap.put(thread.hashCode(), thread);
        thread.start();
        return thread.hashCode();
    }
});

context.property("clearTimer", new JSFunction(context, "clearTimer") {
    public void clearTimer(final Integer hashCode) {
        threadMap.remove(hashCode);
    }
}

```

The one thing you might watch out for is that I don't think the `threadMap` accesses are thread-safe.  You would want to synchronize those probably, or implement a mutex.

**Comment by ericwlange** [comment]
I have added `setTimeout`, `setInterval` and `clearTimer` by default to the raw JavaScript VM.  It is now available in 0.6.1.
